### PR TITLE
Add description to google_service_account

### PR DESCRIPTION
**Pull request by sergei-ivanov**
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
project: added `description` to `google_service_account` data source
```
```release-note:enhancement
project: added `oauth2_client_id` to `google_service_account` data source
```
```release-note:enhancement
project: added `oauth2_client_id` to `google_service_account` resource
```
```release-note:bug
project: fixed update functionality in `google_service_account` resource to allow setting blank `description` and `display_name`
```
```release-note:enhancement
project: added validation to `display_name` in `google_service_account` resource
```

### CODE REVIEW

**Comment by sergei-ivanov** [comment]
Looks like this PR was never picked up and most of the changes have been independently introduced into master. Does it make sense to rebase it on master and merge the remaining changes, or should we just close it?

**Comment by tysen** [comment]
If you still have a need for `oauth2ClientId`, you can rebase and I'll review.

**Comment by chrisst** [comment]
@sergei-ivanov If you are willing to rebase I think this is still valuable.

**Comment by sergei-ivanov** [comment]
Hello @chrisst and @tysen 
Sorry that it took that long. I have just rebased, please take a look.
I have tested it locally as well.

**Comment by sergei-ivanov** [comment]
Updated the changelog in the description according to the wishes of @modular-magician

File: third_party/terraform/resources/resource_google_service_account.go [hunk 1]
```diff
@@ -131,29 +137,44 @@ func resourceGoogleServiceAccountDelete(d *schema.ResourceData, meta interface{}
...
 		return fmt.Errorf("Error retrieving service account %q: %s", d.Id(), err)
 	}
+
+	serviceAccount := &iam.ServiceAccount{
+		Etag: sa.Etag,
+	}
 	updateMask := make([]string, 0)
+
 	if d.HasChange("description") {
+		if v, ok := d.GetOk("description"); ok {
```
Comment (danawillow) [inline third_party/terraform/resources/resource_google_service_account.go:152]: The GCP client libraries don't send empty values unless you ask it to via `ForceSendFields`, so you don't actually need to do the `d.GetOk` check here.

File: third_party/terraform/resources/resource_google_service_account.go [hunk 2]
```diff
@@ -131,29 +137,44 @@ func resourceGoogleServiceAccountDelete(d *schema.ResourceData, meta interface{}
...
+
 	if d.HasChange("display_name") {
+		if v, ok := d.GetOk("display_name"); ok {
+			serviceAccount.DisplayName = v.(string)
+		}
 		updateMask = append(updateMask, "display_name")
 	}
+
+	if len(updateMask) == 0 {
+		return nil
```
Comment (danawillow) [inline third_party/terraform/resources/resource_google_service_account.go:166]: Can you also put something in the debug logs that we found nothing to update? We shouldn't be able to get here unless there was actually a change.

File: third_party/terraform/resources/resource_google_service_account.go (hunk 1 above)
Comment (danawillow) [reply third_party/terraform/resources/resource_google_service_account.go:152]: I'm also not sure why these needed to be moved out of where they were before. What exactly does doing this fix?

File: third_party/terraform/tests/resource_google_service_account_test.go [hunk 3]
```diff
@@ -52,7 +52,7 @@ func TestAccServiceAccount_basic(t *testing.T) {
 			},
 			// The second step updates the service account
 			{
-				Config: testAccServiceAccountBasic(accountId, displayName2, desc2),
+				Config: testAccServiceAccountBasic(accountId, displayName2, description2),
```
Comment (danawillow) [inline third_party/terraform/tests/resource_google_service_account_test.go:55]: Why this change? It's really helpful for us to know that we can set things to empty, since that case often accidentally doesn't work.
